dash_auth/auth.py

```diff
@@ -61,19 +82,43 @@ def before_request_auth():
                 if body["output"] in public_callbacks:
                     return None
 
-                # Check whether the callback has an input using the pathname,
-                # such a callback will be a routing callback and the pathname
-                # should be checked against the public routes
                 pathname = next(
                     (
-                        inp.get("value") for inp in body["inputs"]
+                        inp.get("value")
+                        for inp in body["inputs"]
                         if isinstance(inp, dict)
                         and inp.get("property") == "pathname"
                     ),
                     None,
                 )
-                if pathname and public_routes.test(pathname):
-                    return None
+                if self.page_container:
+                    page_container_test = next(
+                        (
+                            out
+                            for out in body["outputs"]
+                            if isinstance(out, dict)
+                            and out.get("id") == self.page_container
+                            and out.get("property") == "children"
+                        ),
+                        None,
+                    )
+                else:
+                    page_container_test = True
+
+                # Check whether the callback has an input using the pathname,
+                # such a callback will be a routing callback and the pathname
+                # should be checked against the public routes
+                if not self.auth_protect_layouts:
+                    if (
+                        pathname
+                        and page_container_test
+                        and public_routes.test(pathname)
+                    ):
+                        return None
+                else:
+                    # protected by layout
+                    if pathname and page_container_test:
+                        return None
```

Comment:
Let's not do this, this would mean that the user wouldn't be redirected to log in via SSO for instance if you activate your auth_protect_layouts.

Reply:
Hmm, seems to redirect just fine in my test app.

```
import dash
from dash import Dash, html, Input, Output
from dash_auth import OIDCAuth, protected, protected_callback, public_callback
from flask import request, redirect, url_for, session
from flask_login import current_user, LoginManager, login_user, logout_user, UserMixin

app = Dash(__name__, use_pages=True, pages_folder='')

login_manager = LoginManager()
login_manager.init_app(app.server)
class User(UserMixin):
    pass

@login_manager.user_loader
def user_loader(username):
    user = User()
    user.id = username
    return user

def all_login_method(user_info, idp=None):
    if idp:
        session["user"] = user_info
        session["idp"] = idp
        session['user']['groups'] = ['this', 'is', 'a', 'testing', 'nah']
        user = User()
        user.id = user_info['email']
        login_user(user)
    else:
        user = User()
        user.id = user_info.get('user')
        login_user(user)
        session['user'] = {}
        session['user']['groups'] = ['nah', 'testing']
        session['user']['email'] = user_info.get('user')
    return redirect(app.config.get("url_base_pathname") or "/")

def get_groups(**kwargs):
    if kwargs.get('testing'):
        return ['nah']
    return ['testing', 't2', 't3']

restricted_users = ['test', 'testing']
def restricted_user(**kwargs):
    return restricted_users

def layout():
    if request:
        if current_user:
            try:
                return html.Div([
                    html.Div(f"Hello {current_user.id}!"),
                    html.Button(id='change_users', children='change restrictions'),
                    html.Button(id='test', children='you cant use me'),
                    html.A("Logout", href="/logout"),
                    dash.page_container
                ])
            except:
                pass
        if 'user' in session:
            return html.Div([
                html.Div(f"""Hello {session['user'].get('email')}!
                        You have access to these groups: {session['user'].get('groups')}"""),
                html.Button(id='change_users', children='change restrictions'),
                html.Button(id='test', children='you cant use me'),
                html.A("Logout", href="/logout"),
                dash.page_container
            ])
    return html.Div([
        html.Div("Hello world!"),
        html.Button(id='change_users', children='change restrictions'),
        html.Button(id='test', children='you cant use me'),
        html.A("Logout", href="/logout"),
        dash.page_container
    ])

app.layout = layout


def Layout(test=None, **kwargs):
    return html.Div('rawr')
def protected():
    return html.Div('seriously, dont go here')
dash.register_page('test', path_template='/test/<test>', layout=Layout,
                    restricted_users=[''], missing_permissions_output='hahahah'
                   )


@protected_callback(
    Output('test', 'children'),
    Input('test', 'n_clicks'),
    groups=get_groups,
    prevent_initial_call=True,
    group_lookup={'testing': 'rawr'},
    restricted_users=restricted_user
)
def testing(n):
    return 'I was clicked'

@public_callback(
    Output('change_users', 'children'),
    Input('change_users', 'n_clicks'),
    prevent_initial_call=True
)
def alter_restrictions(n):
    global restricted_users
    restricted_users = ['rawr', 'hahaha']
    return 'Users changed'

auth = OIDCAuth(
    app,
    secret_key="aStaticSecretKey!",
    # Set the route at which the user will select the IDP they wish to login with
    idp_selection_route="/login",
    user_groups={'': ['rawr']},
    login_user_callback=all_login_method,
    logout_route='/logout',
    public_routes=[*[f"/static/<path:path>.{ext}" for ext in ['js', 'png', 'gif']]],
    auth_protect_layouts=True,
    auth_protect_layouts_kwargs=dict(missing_permissions_output=html.Div('you cant get me')),
    page_container='_pages_content'
)
auth.register_provider(
    "IDP 1",
    token_endpoint_auth_method="client_secret_post",
    client_id="",
    client_secret="",
    server_metadata_url=f""
)

@app.server.route("/login", methods=["GET", "POST"])
def login_handler():
    if request.method == 'POST':
        form_data = request.form
    else:
        form_data = request.args

    if form_data.get('user') and form_data.get('password'):
        return all_login_method(form_data)

    if form_data.get('IDP 1'):
        return redirect(url_for("oidc_login", idp='IDP 1'))

    return """<div>
        <form method="POST">
            <div>How do you wish to sign in:</div>
            <button type="submit" name="IDP 1" value="true">Microsoft</button>
            <div><input name="user"/>
            <input name="password"/></div>
            <input type="submit" value="Login">
        </form>
    </div>"""


if __name__ == "__main__":
    app.run_server(debug=True)
```

Reply:
It does if you go straight to a protected url / refresh the page. It wouldn't if you navigate there via a callback. For instance if `/` is public and links to `/private` which isn't.